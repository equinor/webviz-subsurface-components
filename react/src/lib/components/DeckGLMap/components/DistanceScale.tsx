import React from "react";
// eslint-disable-next-line @typescript-eslint/no-var-requires
const convert = require("convert-units");

interface scaleProps {
    // Needed the zoom value to calculate width in units
    zoom: number;
    // Scale increment value
    incrementValue: number;
    // Scale bar width in pixels per unit value
    widthPerUnit: number;
    // positioning the scale ruler based on x and y values
    position: number[];
    // default unit for the scale ruler
    scaleUnit: string;
}

const roundToStep = function (num: number, step: number) {
    return Math.floor(num / step + 0.5) * step;
};

const DistanceScale: React.FC<scaleProps> = ({
    zoom,
    incrementValue,
    widthPerUnit,
    position,
    scaleUnit,
}: scaleProps) => {
    const [rulerWidth, setRulerWidth] = React.useState<number>(0);
    const widthInUnits = widthPerUnit / Math.pow(2, zoom);
    const scaleRulerStyle = {
        width: rulerWidth,
        height: "4px",
        border: "2px solid gray",
        borderTop: "none",
        display: "inline-block",
        marginLeft: "3px",
        marginRight: "3px",
        right: 0,
        bottom: 0,
    };

    const scaleValue =
        widthInUnits < incrementValue
            ? Math.round(widthInUnits)
            : roundToStep(widthInUnits, incrementValue);

    const convertedUnit = convert(scaleValue).from(scaleUnit).toBest().unit;
    const convertedValue = convert(scaleValue).from(scaleUnit).toBest().val;

    React.useEffect(() => {
        setRulerWidth(scaleValue * Math.pow(2, zoom));
    }, [zoom]);

    return (
        <div
            style={{
                bottom: position[0],
                left: position[1],
                position: "relative",
            }}
        >
            <label>{convertedValue.toFixed(0)}</label>
            {convertedUnit}
            <div style={scaleRulerStyle}></div>
        </div>
    );
};

DistanceScale.defaultProps = {
    incrementValue: 100,
    widthPerUnit: 100,
    position: [10, 10],
    scaleUnit: "m",
};

export default DistanceScale;
