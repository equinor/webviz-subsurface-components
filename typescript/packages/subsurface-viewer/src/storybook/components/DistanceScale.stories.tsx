import type { Meta, StoryObj } from "@storybook/react";

import DistanceScale from "../../components/DistanceScale";

const stories: Meta = {
    component: DistanceScale,
    title: "SubsurfaceViewer / Components / DistanceScale",
};
export default stories;

const darkModeStyle = {
    color: "white",
};

export const LightMode: StoryObj<typeof DistanceScale> = {};

export const DarkMode: StoryObj<typeof DistanceScale> = {
    args: {
        style: darkModeStyle,
    },
    parameters: {
        backgrounds: { default: "dark" },
    },
};
