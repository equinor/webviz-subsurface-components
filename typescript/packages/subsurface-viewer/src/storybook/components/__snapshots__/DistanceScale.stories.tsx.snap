// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubsurfaceViewer / Components / DistanceScale DarkMode smoke-test 1`] = `
<div style="position: absolute; color: white;">
  <label style="color: white;">
    800m
  </label>
  <div style="width: 100px; height: 4px; border-top: none; border-right: 2px solid; border-bottom: 2px solid; border-left: 2px solid; border-image: initial; display: inline-block; margin-left: 3px;">
  </div>
</div>
`;

exports[`SubsurfaceViewer / Components / DistanceScale LightMode smoke-test 1`] = `
<div style="position: absolute;">
  <label>
    800m
  </label>
  <div style="width: 100px; height: 4px; border-top: none; border-right: 2px solid; border-bottom: 2px solid; border-left: 2px solid; border-image: initial; display: inline-block; margin-left: 3px;">
  </div>
</div>
`;
